<a href="https://colab.research.google.com/github/harishmuh/Python-for-Data-Science-Analysis/blob/main/Pandas_II_Exercise.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Exercise - Pandas II**

In [1]:
# Importing libraries
import numpy as np
import pandas as pd

**1. Load the Salaries.csv dataset, then save it in a variable named df**

You can download the dataset from [Here](https://github.com/harishmuh/Python-for-Data-Science-Analysis/blob/main/datasets/Salaries.csv) or use this [Link](https://raw.githubusercontent.com/harishmuh/Python-for-Data-Science-Analysis/refs/heads/main/datasets/Salaries.csv) to obtain the dataset directly from GitHub.

In [2]:
# Loading dataset
url = 'https://raw.githubusercontent.com/harishmuh/Python-for-Data-Science-Analysis/refs/heads/main/datasets/Salaries.csv'
df = pd.read_csv(url)

**2. Display the first 10 rows of the dataframe df**

In [3]:
df.head(10)

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Notes,Agency,Status
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,NaN,567595.43,567595.43,2011,NaN,San Francisco,NaN
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,NaN,538909.28,538909.28,2011,NaN,San Francisco,NaN
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,NaN,335279.91,335279.91,2011,NaN,San Francisco,NaN
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,NaN,332343.61,332343.61,2011,NaN,San Francisco,NaN
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,NaN,326373.19,326373.19,2011,NaN,San Francisco,NaN
5,6,DAVID SULLIVAN,ASSISTANT DEPUTY CHIEF II,118602.00,8601.00,189082.74,NaN,316285.74,316285.74,2011,NaN,San Francisco,NaN
6,7,ALSON LEE,"BATTALION CHIEF, (FIRE DEPARTMENT)",92492.01,89062.90,134426.14,NaN,315981.05,315981.05,2011,NaN,San Francisco,NaN
7,8,DAVID KUSHNER,DEPUTY DIRECTOR OF INVESTMENTS,256576.96,0.00,51322.50,NaN,307899.46,307899.46,2011,NaN,San Francisco,NaN
8,9,MICHAEL MORRIS,"BATTALION CHIEF, (FIRE DEPARTMENT)",176932.64,86362.68,40132.23,NaN,303427.55,303427.55,2011,NaN,San Francisco,NaN
9,10,JOANNE HAYES-WHITE,"CHIEF OF DEPARTMENT, (FIRE DEPARTMENT)",285262.00,0.00,17115.73,NaN,302377.73,302377.73,2011,NaN,San Francisco,NaN


**3. Display dataframe information regarding column names, the amount of data filled in, and the data types**

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  object 
 2   JobTitle          148654 non-null  object 
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  object 
 12  Status            0 non-null       float64
dtypes: float64(8), int64(2), object(3)
memory usage: 14.7+ MB


**4. Provide an explanation regarding the dataframe information**

Answer:
- The DataFrame consists of 148,654 rows and 13 columns.
- 2 columns are integers and 8 columns are floats.
- 3 columns are strings (objects).
- 6 columns have null data: BasePay, OvertimePay, OtherPay, Benefits, Notes, and Status.

**5. Display the number of rows from the dataframe df**

In [5]:
df.shape[0]

148654

**6. Display the number of missing values ​​in the dataframe**

In [7]:
df.isna().sum()[df.isna().sum()>0].to_frame('missing values')

,missing values
BasePay,609
OvertimePay,4
OtherPay,4
Benefits,36163
Notes,148654
Status,148654


**7. Display the number of missing values ​​in the BasePay day OvertimePay column**

In [9]:
df[['BasePay', 'OvertimePay']].isna().sum().to_frame('missing values')

,missing values
BasePay,609
OvertimePay,4


**8. Display employee data whose OtherPay is missing (NaN)**

In [10]:
df[df['OtherPay'].isna()]

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Notes,Agency,Status
148646,148647,Not provided,Not provided,NaN,NaN,NaN,NaN,0.0,0.0,2014,NaN,San Francisco,NaN
148650,148651,Not provided,Not provided,NaN,NaN,NaN,NaN,0.0,0.0,2014,NaN,San Francisco,NaN
148651,148652,Not provided,Not provided,NaN,NaN,NaN,NaN,0.0,0.0,2014,NaN,San Francisco,NaN
148652,148653,Not provided,Not provided,NaN,NaN,NaN,NaN,0.0,0.0,2014,NaN,San Francisco,NaN


**9. Show statistics for numeric columns**

In [11]:
df.describe()

,Id,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Notes,Status
count,148654.000000,148045.000000,148650.000000,148650.000000,112491.000000,148654.000000,148654.000000,148654.000000,0.0,0.0
mean,74327.500000,66325.448840,5066.059886,3648.767297,25007.893151,74768.321972,93692.554811,2012.522643,NaN,NaN
std,42912.857795,42764.635495,11454.380559,8056.601866,15402.215858,50517.005274,62793.533483,1.117538,NaN,NaN
min,1.000000,-166.010000,-0.010000,-7058.590000,-33.890000,-618.130000,-618.130000,2011.000000,NaN,NaN
25%,37164.250000,33588.200000,0.000000,0.000000,11535.395000,36168.995000,44065.650000,2012.000000,NaN,NaN
50%,74327.500000,65007.450000,0.000000,811.270000,28628.620000,71426.610000,92404.090000,2013.000000,NaN,NaN
75%,111490.750000,94691.050000,4658.175000,4236.065000,35566.855000,105839.135000,132876.450000,2014.000000,NaN,NaN
max,148654.000000,319275.010000,245131.880000,400184.250000,96570.660000,567595.430000,567595.430000,2014.000000,NaN,NaN


**10. Provide an explanation regarding the statistical information that has been obtained**

In general, we can see that:
* Based on **Year** data, the years of occurrence ranged from 2011 to 2014.
* **BasePay** ranged from -166 to 319,275 with an average of 66,325.
* 75% of employees had **OvertimePay** below 4,658, but some employees had **OvertimePay** up to 2,45,131.
* Some employees had **OtherPay** reaching 400,184, but some employees had negative **OtherPay**.

**11. Show statistics for categorical columns**

In [12]:
df.describe(include='object')

,EmployeeName,JobTitle,Agency
count,148654,148654,148654
unique,110811,2159,1
top,Kevin Lee,Transit Operator,San Francisco
freq,13,7036,148654


**12. Provide an explanation regarding the statistical information that has been obtained**

In [13]:
df[df['EmployeeName']=='Rosemary Lee'][['EmployeeName',	'JobTitle']]

,EmployeeName,JobTitle
36497,Rosemary Lee,Nurse Manager


**13. Show employee named Rosemary Lee and her job**

In [14]:
df[df['EmployeeName']=='Rosemary Lee'][['EmployeeName',	'JobTitle']]

,EmployeeName,JobTitle
36497,Rosemary Lee,Nurse Manager


**14. Display the name of the employee with the largest TotalPayBenefits**

In [15]:
df[df['TotalPayBenefits']==df['TotalPayBenefits'].max()]['EmployeeName'][0]

'NATHANIEL FORD'

In [16]:
df.loc[df['TotalPayBenefits'].idxmax(), 'EmployeeName']

'NATHANIEL FORD'

**15. Display dataset sorted from oldest year, then from highest to lowest BasePay**

In [17]:
df.sort_values(['Year', 'BasePay'], ascending=[True, False])

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Notes,Agency,Status
12,13,EDWARD HARRINGTON,EXECUTIVE CONTRACT EMPLOYEE,294580.02,0.0,0.00,NaN,294580.02,294580.02,2011,NaN,San Francisco,NaN
9,10,JOANNE HAYES-WHITE,"CHIEF OF DEPARTMENT, (FIRE DEPARTMENT)",285262.00,0.0,17115.73,NaN,302377.73,302377.73,2011,NaN,San Francisco,NaN
13,14,JOHN MARTIN,DEPARTMENT HEAD V,271329.03,0.0,21342.59,NaN,292671.62,292671.62,2011,NaN,San Francisco,NaN
16,17,AMY HART,DEPARTMENT HEAD V,268604.57,0.0,16115.86,NaN,284720.43,284720.43,2011,NaN,San Francisco,NaN
28,29,DENISE SCHMITT,DEPUTY CHIEF III (POLICE DEPARTMENT),261717.60,0.0,2357.00,NaN,264074.60,264074.60,2011,NaN,San Francisco,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...
148653,148654,Joe Lopez,"Counselor, Log Cabin Ranch",0.00,0.0,-618.13,0.0,-618.13,-618.13,2014,NaN,San Francisco,NaN
148646,148647,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,NaN,San Francisco,NaN
148650,148651,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,NaN,San Francisco,NaN
148651,148652,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,NaN,San Francisco,NaN


**16. Show employees with BasePay above 300 thousand dollars and OvertimePay not 0 dollars**

In [18]:
df.loc[(df['BasePay']>300000) & (df['OvertimePay']!=0), ['EmployeeName', 'BasePay', 'OvertimePay']]

,EmployeeName,BasePay,OvertimePay
110532,Amy P Hart,318835.49,10712.95


In [19]:
df[(df['BasePay'] > 300000) & (df['OvertimePay'] != 0)][['EmployeeName', 'BasePay', 'OvertimePay']]

,EmployeeName,BasePay,OvertimePay
110532,Amy P Hart,318835.49,10712.95


**17. Show BasePay rate for the first 10,000 data**

In [20]:
df.head(10000)['BasePay'].mean()

np.float64(112236.757339)

**18. Show the number of unique JobTitles**

In [21]:
df['JobTitle'].nunique()

2159

**19. Show 7 jobs with the highest frequency**

In [22]:
# Metod 1
df['JobTitle'].value_counts().to_frame().reset_index().rename(columns={'index':'JobTitle', 'JobTitle':'count'}).head(7)

,count,count
0,Transit Operator,7036
1,Special Nurse,4389
2,Registered Nurse,3736
3,Public Svc Aide-Public Works,2518
4,Police Officer 3,2421
5,Custodian,2418
6,TRANSIT OPERATOR,2388


In [ ]:
# Method 2
dict_job = {}

for title in df['JobTitle']:
    if title not in dict_job.keys():
        dict_job[title] = 1
    else:
        dict_job[title] += 1

pd.DataFrame({
    'JobTitle' : dict_job.keys(),
    'Frekuensi' : dict_job.values()
}).sort_values('Frekuensi', ascending=False).head(7)

**20. How many JobTitles appeared only once in 2013**

In [23]:
sum(df[df['Year']==2013]['JobTitle'].value_counts()==1)

202